### PR TITLE
8310606: Fix signed integer overflow, part 3

### PR DESCRIPTION
**Pull request by dean-long**
Most of these changes are straight-forward, but these could use some explaining:

AllocatePrefetchDistance was accessed before it's dynamic constraint function got run.  I changed it to use static constraints.

RangeCheckEliminator::update_bound(): for constant bounds with meaningless values, bail out.  For X + c bounds, use Java wrap-around semantics.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8310606](https://bugs.openjdk.org/browse/JDK-8310606): Fix signed integer overflow, part 3 (**Sub-task** - P4)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [85868499](https://git.openjdk.org/jdk/pull/14619/files/8586849944055d28d9e1f67c66a79831ca2c9ec3)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [85868499](https://git.openjdk.org/jdk/pull/14619/files/8586849944055d28d9e1f67c66a79831ca2c9ec3)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14619/head:pull/14619` \
`$ git checkout pull/14619`

Update a local copy of the PR: \
`$ git checkout pull/14619` \
`$ git pull https://git.openjdk.org/jdk.git pull/14619/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14619`

View PR using the GUI difftool: \
`$ git pr show -t 14619`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14619.diff">https://git.openjdk.org/jdk/pull/14619.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14619#issuecomment-1603241993)

### CODE REVIEW

**Review by vnkozlov — APPROVED** [review]
Looks good.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

File: src/hotspot/cpu/aarch64/vm_version_aarch64.cpp [hunk 1]
```diff
@@ -79,7 +79,7 @@ void VM_Version::initialize() {
   int dcache_line = VM_Version::dcache_line_size();
 
   // Limit AllocatePrefetchDistance so that it does not exceed the
-  // constraint in AllocatePrefetchDistanceConstraintFunc.
+  // constraint.
   if (FLAG_IS_DEFAULT(AllocatePrefetchDistance))
     FLAG_SET_DEFAULT(AllocatePrefetchDistance, MIN2(512, 3*dcache_line));
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/vm_version_aarch64.cpp:84]: Looks reasonable enough, but I think the magic constant 512 deserves a comment.

File: src/hotspot/share/jfr/leakprofiler/utilities/granularTimer.cpp [hunk 2]
```diff
@@ -36,8 +36,8 @@ void GranularTimer::start(jlong duration_ticks, long granularity) {
   _granularity = granularity;
   _counter = granularity;
   _start_time_ticks = JfrTicks::now();
-  const jlong end_time_ticks = _start_time_ticks.value() + duration_ticks;
-  _finish_time_ticks = end_time_ticks < 0 ? JfrTicks(max_jlong) : JfrTicks(end_time_ticks);
+  const julong end_time_ticks = (julong)_start_time_ticks.value() + (julong)duration_ticks;
+  _finish_time_ticks = end_time_ticks > (julong)max_jlong ? JfrTicks(max_jlong) : JfrTicks(end_time_ticks);
   _finished = _finish_time_ticks == _start_time_ticks;
```
Comment (theRealAph) [inline src/hotspot/share/jfr/leakprofiler/utilities/granularTimer.cpp:41]: Isn't there a much better way to fix this?

Comment (theRealAph) [reply src/hotspot/share/jfr/leakprofiler/utilities/granularTimer.cpp:41]: If you change 

`if (JfrTicks::now() > _finish_time_ticks) {`

to

`if (JfrTicks::now() - _finish_time_ticks > 0) {`

you get natural wrap-around semantics. As long as the total interval is less than half the total julong range. And you don't need to worry about `max_jlong`, etc.

Comment (dean-long) [reply src/hotspot/share/jfr/leakprofiler/utilities/granularTimer.cpp:41]: The problem here is Java code is passing a value of max_jlong for duration_ticks.  I don't see how changing the logic (in is_finished()?) helps with that.

**Pull request closed without merging**